extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java

```diff
@@ -119,4 +136,26 @@ void registerAdditionalBeans(BuildProducer<AdditionalBeanBuildItem> beans) {
         beans.produce(AdditionalBeanBuildItem.unremovableOf(IdentityProviderManagerCreator.class));
         beans.produce(AdditionalBeanBuildItem.unremovableOf(SecurityIdentityProxy.class));
     }
+
+    @BuildStep
+    void validateAnnotatedMethods(ValidationPhaseBuildItem validationPhase,
+            BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> errors) {
+
+        AnnotationStore annotationStore = validationPhase.getContext().get(BuildExtension.Key.ANNOTATION_STORE);
+        for (BeanInfo bean : validationPhase.getContext().get(BuildExtension.Key.BEANS)) {
+            if (bean.isClassBean()) {
+                // Ensure that all security annotations are placed on non final methods
+                // this is done because Arc only warns about final methods not being interceptable
+                for (MethodInfo method : bean.getTarget().get().asClass().methods()) {
```

Comment:
Inherited methods are not processed. Also iterating over all methods of all beans could be quite time-consuming (for large deployments). I'd rather move the validation to `io.quarkus.arc.processor.BeanInfo.initInterceptedMethods()` or somewhere "nearby".

Reply:
I pondered that option as well but went with the current solution because it seemed cleaner since it wouldn't involve baking specific security annotations inside the general Arc code.
I can certainly move things inside `initInterceptedMethods` but I assume that the security annotations will have to be hard coded - or at best have some kind of build item that is dedicated to this case and will propagate the information down the Arc layers.
What would be your preference?

Reply:
It might be a stupid question but why don't we throw an error for any interceptor binding annotation on final methods? I mean it won't work and could have very bad consequences. Security is one of the occurrences of this problem but there might be other.

Reply:
That's fine with me TBH. I just didn't do it because it was mentioned in the issue that we might not want that as a default.

Reply:
I think that we should modify [BeanDeploymentValidator#skipValidation()](https://github.com/quarkusio/quarkus/blob/master/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeploymentValidator.java#L31) to handle similar use cases...

Reply:
@mkouba the problem with that approach is that at the point those are executed, we have long lost the interceptor information.

Reply:
> It might be a stupid question ...

We could and we should (according to the spec) but we had some user requests to ignore those methods and in CDI 2.0 we introduced things like `ProcessBeanAttributes#ignoreFinalMethods()` and `InterceptionFactory#ignoreFinalMethods()`, and in Weld we have a [special config property](https://docs.jboss.org/weld/reference/latest/en-US/html/configure.html#_proxying_classes_with_final_methods) to ignore final methods on some classes.

> but I assume that the security annotations will have to be hard coded...

I'd like to propose some API change to the `BeanDeploymentValidator` API. Hard-coded values are definitely a "no go" ;-).

Reply:
OK cool. Waiting for your ping then :)

Reply:
I think the default behaviour should be to fail on final methods. If users want to turn off that behaviour that is fine, but any mistakes are then on them.

Reply:
I tend to agree although I don't have the experience you folks have on this.
 
From my perspective it does make sense to fail when an interceptor can't be created since that creates the least surprise and the reason for the failure it absolutely clear. Heck if we wanted to we could even do some extra work to tell if the class is a Kotlin class and let the user know that then need to `open` the method.

So it probably makes sense to have a configuration option that default to failing instead of coming up with a more elaborate Arc API for validating.

But like I said before, I'm cool with whatever you folks think is best. Just let me know and I'll update the PR :)

Reply:
Really I think that automatically removing the final modifier is probably the most user friendly approach, especially for Kotlin applications.

Reply:
@stuartwdouglas are you thinking for this interceptor use case or for all types of issues we might encounter with Kotlin?

Reply:
I was just thinking for the interceptor use case.

Reply:
OK, I'll do it and we can see what @mkouba thinks :)

Reply:
Here you go: https://github.com/quarkusio/quarkus/pull/5104